src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp

```diff
@@ -16,33 +23,112 @@
 /// \cond
 namespace NewtonianEuler {
 namespace Sources {
+template <size_t Dim>
+VortexPerturbation<Dim>::VortexPerturbation(
+    const double adiabatic_index, const double perturbation_amplitude,
+    const std::array<double, Dim>& vortex_center,
+    const std::array<double, Dim>& vortex_mean_velocity,
+    const double vortex_strength) noexcept
+    : adiabatic_index_(adiabatic_index),
+      perturbation_amplitude_(perturbation_amplitude),
+      vortex_center_(vortex_center),
+      vortex_mean_velocity_(vortex_mean_velocity),
+      vortex_strength_(vortex_strength) {}
 
-void VortexPerturbation::apply(
-    const gsl::not_null<Scalar<DataVector>*> mass_density_source,
-    const gsl::not_null<tnsr::I<DataVector, 3>*> momentum_density_source,
-    const gsl::not_null<Scalar<DataVector>*> energy_density_source,
-    const Scalar<DataVector>& vortex_mass_density,
-    const tnsr::I<DataVector, 3>& vortex_momentum_density,
-    const Scalar<DataVector>& vortex_energy_density,
-    const Scalar<DataVector>& vortex_pressure,
-    const Scalar<DataVector>& vortex_velocity_z,
-    const Scalar<DataVector>& dz_vortex_velocity_z) noexcept {
-  get(*mass_density_source) =
-      get(vortex_mass_density) * get(dz_vortex_velocity_z);
+template <size_t Dim>
+void VortexPerturbation<Dim>::pup(PUP::er& p) noexcept {
+  p | adiabatic_index_;
+  p | perturbation_amplitude_;
+  p | vortex_center_;
+  p | vortex_mean_velocity_;
+  p | vortex_strength_;
+}
+
+template <>
+void VortexPerturbation<2>::apply() const noexcept {}
+
+template <>
+void VortexPerturbation<3>::apply(
+    const gsl::not_null<Scalar<DataVector>*> source_mass_density_cons,
+    const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> source_energy_density,
+    const tnsr::I<DataVector, 3>& x, const double time) const noexcept {
+  Solutions::IsentropicVortex<3> vortex(
```

Comment:
pass the solution as an argument so it is not recomputed everytime... (see comment in hpp file)